src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -1855,17 +2033,89 @@ bool isRefEscape
             ImmutableArray<RefKind> argRefKindsOpt,
             ImmutableArray<int> argsToParamsOpt,
             bool isRefEscape,
-            bool checkingMethodArgumentsMustMatch,
             bool ignoreArglistRefKinds,
-            ArrayBuilder<(ParameterSymbol? Parameter, BoundExpression Argument, bool IsRefEscape)> result)
+            ArrayBuilder<EscapeValue> escapeValues)
+        {
+            GetFilteredArguments(
+                symbol,
+                receiver,
+                parameters,
+                argsOpt,
+                argRefKindsOpt,
+                argsToParamsOpt,
+                ignoreArglistRefKinds,
+                mixableArguments: null,
+                escapeValues);
+
+            bool includeRefScopes = shouldIncludeRefScopes();
+            int index = 0;
+            while (index < escapeValues.Count)
+            {
+                var tuple = escapeValues[index];
+                if (!includeRefScopes && tuple.IsRefEscape)
+                {
+                    escapeValues.RemoveAt(index);
+                }
+                else
+                {
+                    index++;
+                }
+            }
+
+            bool shouldIncludeRefScopes()
+            {
+                switch (symbol)
+                {
+                    case MethodSymbol method:
+                        if (method.MethodKind == MethodKind.Constructor)
+                        {
+                            return method.ContainingType.IsRefLikeType;
+                        }
+
+                        if (method.ReturnType.IsRefLikeType)
+                        {
+                            return true;
+                        }
+
+                        return method.RefKind != RefKind.None && isRefEscape;
+                    case PropertySymbol property:
+                        if (property.Type.IsRefLikeType)
+                        {
+                            return true;
+                        }
+
+                        return property.RefKind != RefKind.None && isRefEscape;
+
+                    default:
+                        return false;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns the set of arguments to an invocation that impact ref analysis. This will 
+        /// filter out everything that could never meaningfully contribute to ref analysis. For
+        /// example it will remove anything which corresponds to scope parameters. For ref
+        /// parameters it will place in two values representing both the RSTS and STS
+        /// </summary>
+        private void GetFilteredArguments(
```

Comment:
>GetFilteredArguments [](http://example.com/codeflow?start=21&length=20)

Consider using a more specific name so it's clear this is used for escape analysis (and with updated escape rules).